src/WeekView/WeekView.js

```diff
@@ -254,99 +280,184 @@ export default class WeekView extends Component {
-      const prependNeeded = firstViewablePage - targetIndex;
+      const reIndexedTargetPage = newPages.length - PAGES_OFFSET;
 
-      newState.initialDates = [
-        ...this.buildPages(initialDates[0], prependNeeded, false),
-        ...initialDates,
-      ];
-      targetIndex = PAGES_OFFSET;
-
-      newStateCallback = () => setTimeout(() => scrollTo(targetIndex), 0);
-    } else if (targetIndex > lastViewablePage) {
-      const appendNeeded = targetIndex - lastViewablePage;
-      newState.initialDates = [
-        ...initialDates,
-        ...this.buildPages(
-          initialDates[initialDates.length - 1],
-          appendNeeded,
-          true,
-        ),
+      return [
+        reIndexedTargetPage,
+        {
+          initialDates: newPages,
+          currentMoment: date2State(newPages[reIndexedTargetPage]),
+        },
       ];
+    }
+    return [
+      targetPageIndex,
+      {
+        currentMoment: date2State(oldPages[targetPageIndex]),
+      },
+    ];
+  };
+
+  /**
+   * Computes the left-offset displayed in the current date.
+   *
+   * Helper method used in goToPageIndex()
+   * */
+  getCurrentDayOffset = () => {
+    const { initialDates, currentMoment } = this.state;
 
-      targetIndex = newState.initialDates.length - PAGES_OFFSET;
+    return moment(currentMoment).diff(
+      initialDates[this.currentPageIndex],
+      'day',
+    );
+  };
 
-      newStateCallback = () => setTimeout(() => scrollTo(targetIndex), 0);
-    } else {
-      scrollTo(targetIndex);
+  /**
+   * Navigates the view to a pageIndex and (optional) dayOffset.
+   *
+   * Adds more pages (if necessary), scrolls the List to the new index,
+   * and updates this.currentPageIndex.
+   *
+   * @param {Number} targetPageIndex between (-infinity, infinity) indicating target page.
+   * @param {Number} targetDayOffset day offset inside a page.
+   *     Only used if allowScrollByOneDay is true.
+   */
+  goToPageIndex = (targetPageIndex, targetDayOffset, options = {}) => {
+    const { allowScrollByOneDay } = this.props;
+    if (targetPageIndex === this.currentPageIndex && !allowScrollByOneDay) {
+      // If allowScrollByOneDay is false, cannot scroll through offsets
+      return;
     }
 
-    newState.currentMoment = moment(initialDates[targetIndex]).toDate();
-    this.setState(newState, newStateCallback);
+    const dayOffset =
+      !allowScrollByOneDay || targetDayOffset == null
+        ? this.getCurrentDayOffset()
+        : targetDayOffset;
+
+    const viewOffset =
+      targetDayOffset == null
+        ? undefined
+        : VIEW_OFFSET_SIGN * this.dimensions.dayWidth * dayOffset;
+
+    const [moveToIndex, newState] = this.computeParamsToGoToIndex(
+      targetPageIndex,
+      targetDayOffset || 0,
+    );
+
+    const { animated = true } = options || {};
+
+    this.setState(newState, () =>
+      // setTimeout is used to force calling scroll after UI is updated
+      setTimeout(() => {
```

Comment:
if the callback is intented to run after UI update maybe it's better [to use `requestAnimationFrame`](https://reactnative.dev/docs/timers#timers)?

Reply:
I think the best solution should be [runAfterInteractions](https://reactnative.dev/docs/timers#interactionmanager), but it has not worked in the previous versions, not sure why :confused:

I might try these again

Reply:
yeah, let's check again later